File: libsrreal/diffpy/srreal/ScatteringFactorTable.cpp
/*****************************************************************************
*
* diffpy.srreal     by DANSE Diffraction group
*                   Simon J. L. Billinge
*                   (c) 2009 Trustees of the Columbia University
*                   in the City of New York.  All rights reserved.
*
* File coded by:    Pavol Juhas
*
* See AUTHORS.txt for a list of people who contributed.
* See LICENSE.txt for license information.
*
******************************************************************************
*
* class ScatteringFactorTable -- base class for looking up scattering factors
*
* $Id$
*
*****************************************************************************/

#include <diffpy/srreal/ScatteringFactorTable.hpp>
#include <diffpy/ClassRegistry.hpp>

using namespace std;
using diffpy::ClassRegistry;

namespace diffpy {
namespace srreal {

// class ScatteringFactorTable -----------------------------------------------

// public methods

const double& ScatteringFactorTable::lookup(const string& smbl) const
{
    using namespace std;
    map<string, double>::const_iterator isft;
    isft = mtable.find(smbl);
    if (isft == mtable.end())
    {
        double value = this->fetch(smbl);
        mtable[smbl] = value;
        isft = mtable.find(smbl);
    }
    return isft->second;
}


void ScatteringFactorTable::setCustom(const string& smbl, double value)
{
    mtable[smbl] = value;
}


void ScatteringFactorTable::resetCustom(const string& smbl)
{
    mtable.erase(smbl);
}


void ScatteringFactorTable::resetAll()
{
    mtable.clear();
}

// Factory Functions ---------------------------------------------------------

ScatteringFactorTable* createScatteringFactorTable(const string& tp)
{
    return ClassRegistry<ScatteringFactorTable>::create(tp);
}


bool registerScatteringFactorTable(const ScatteringFactorTable& ref)
{
    return ClassRegistry<ScatteringFactorTable>::add(ref);
}


bool aliasScatteringFactorTable(const string& tp, const string& al)
{
    return ClassRegistry<ScatteringFactorTable>::alias(tp, al);
}


set<string> getScatteringFactorTableTypes()
{
    return ClassRegistry<ScatteringFactorTable>::getTypes();
}

}   // namespace srreal
}   // namespace diffpy

// End of file

File: diffpy/srreal/scatteringfactortable.py
#!/usr/bin/env python
##############################################################################
#
# diffpy.srreal     by DANSE Diffraction group
#                   Simon J. L. Billinge
#                   (c) 2010 The Trustees of Columbia University
#                   in the City of New York.  All rights reserved.
#
# File coded by:    Pavol Juhas
#
# See AUTHORS.txt for a list of people who contributed.
# See LICENSE_DANSE.txt for license information.
#
##############################################################################


"""\
class ScatteringFactorTable -- scattering factors for atoms, ions and isotopes.
"""


# exported items, these also makes them show in pydoc.
__all__ = ['ScatteringFactorTable',
        'SFTXray', 'SFTElectron', 'SFTNeutron', 'SFTElectronNumber',
        'SFAverage']

from diffpy.srreal.srreal_ext import ScatteringFactorTable
from diffpy.srreal.srreal_ext import SFTXray
from diffpy.srreal.srreal_ext import SFTElectron
from diffpy.srreal.srreal_ext import SFTNeutron
from diffpy.srreal.srreal_ext import SFTElectronNumber
from diffpy.srreal.wraputils import _pickle_getstate, _pickle_setstate
from diffpy.srreal.sfaverage import SFAverage

# Pickling Support -----------------------------------------------------------

def _sft_create(owner):
    return owner.scatteringfactortable

def _sft_reduce(self):
    from diffpy.srreal.srreal_ext import ScatteringFactorTableOwner
    owner = ScatteringFactorTableOwner()
    owner.scatteringfactortable = self
    args = (owner,)
    rv = (_sft_create, args)
    return rv

def _sft_reduce_with_state(self):
    rv = _sft_reduce(self) + (self.__getstate__(),)
    return rv

# inject pickle methods to ScatteringFactorTable

ScatteringFactorTable.__reduce__ = _sft_reduce_with_state
ScatteringFactorTable.__getstate__ = _pickle_getstate
ScatteringFactorTable.__setstate__ = _pickle_setstate

SFTXray.__reduce__ = _sft_reduce
SFTElectron.__reduce__ = _sft_reduce
SFTNeutron.__reduce__ = _sft_reduce
SFTElectronNumber.__reduce__ = _sft_reduce

# End of file
